statsmodels/formula/_manager.py

```diff
@@ -0,0 +1,452 @@
+    ):
+        """
+        Get the linear constraints from the constraints and variable names.
+
+        Parameters
+        ----------
+        constraints
+        variable_names
+
+        Returns
+        -------
+        LinearConstraintValues
+            The constraint matrix, constraint values, and variable names.
+        """
+        if self._engine == "patsy":
+            from patsy.design_info import DesignInfo
+
+            lc = DesignInfo(variable_names).linear_constraint(constraints)
+            return LinearConstraintValues(
+                constraint_matrix=lc.coefs,
+                constraint_values=lc.constants,
+                variable_names=lc.variable_names,
+            )
+        else:  # self._engine == "formulaic"
+
+            # Handle list of constraints, which is not supported by formulaic
+            if isinstance(constraints, list):
+                if not all(isinstance(c, str) for c in constraints):
+                    raise ValueError(
+                        "All constraints must be strings when passed as a list."
+                    )
+                _constraints = ", ".join(str(v) for v in constraints)
+            else:
+                _constraints = constraints
+            lc_f = formulaic.utils.constraints.LinearConstraints.from_spec(
+                _constraints, variable_names=list(variable_names)
+            )
+            return LinearConstraintValues(
+                constraint_matrix=lc_f.constraint_matrix,
+                constraint_values=np.atleast_2d(lc_f.constraint_values).T,
+                variable_names=lc_f.variable_names,
+            )
+
+    def get_empty_eval_env(self):
+        """
+        Get an empty evaluation environment.
+
+        Returns
+        -------
+        {EvalEnvironment, dict}
+            A formula-engine-dependent empty evaluation environment.
+        """
+        if self._engine == "patsy":
+            from patsy.eval import EvalEnvironment
+
+            return EvalEnvironment({})
+        else:
+            return {}
+
+    @property
+    def intercept_term(self):
+        """
+        Get the formula-engine-specific intercept term.
+
+        Returns
+        -------
+        Term
+            The intercept term.
+        """
+        if self._engine == "patsy":
+            from patsy.desc import INTERCEPT
+
+            return INTERCEPT
+        else:
+            from formulaic.parser.types.factor import Factor
+            from formulaic.parser.types.term import Term
+
+            return Term((Factor("1"),))
+
+    def remove_intercept(self, terms):
+        """
+        Remove intercept from Patsy terms.
+
+        Parameters
+        ----------
+        terms : list[Term]
+            A list of terms that might have an intercept
+
+        Returns
+        -------
+        list[Term]
+            The terms with the intercept removed, if present.
+        """
+        intercept = self.intercept_term
```

Comment:
question: removing the intercept term does not mean that generated matrices will not span the intercept; both patsy and formulaic will, in the absence of an intercept, expand categorical factors to span the intercept. How is that being handled?

Reply:
TBH I'm not that worried about this - it is mostly to deal with code that exists that someone else wrote a long time ago. It is only used in 1 place for anova, and then only to loop over columns that aren't the intercept. It isn't used to construct a model matrix.

Reply:
@matthewwardrop 

In models where we cannot have an intercept like OrderedModel, we require users when using formulas to use an explicit intercept that we remove internally.
That means that the remaining variables are by construction not spanning the intercept.
(patsy does not have an option to exclude explicit and implicit intercepts)

Reply:
Thanks for the context @josef-pkt .

Reply:
(fwiw, if it ever became sufficiently desirous to have this built into a formula parser, I'd be open to considering how best achieve it; but this seems like a reasonable approach for now)